### DISCUSSION TITLE
v32.0-alpha1

### DISCUSSION THREAD

**Discussion by westnordost** [streetcomplete/StreetComplete#2834] [Release Discussions]
<p>
  The grant from the <a href="https://prototypefund.de/en/project/streetcomplete/">Prototype Fund</a>
  ended in February, however, there is one last big mega-update I can delight you with before development
  on this app must necessarily wind down:<br/>
  It may not look like a lot, but I basically replaced the whole architecture of the app (#2506, ...)
  to do this. Special thanks to @FloEdelmann for helping with the refactor.
  </p>
  <h3>🔙️ Superpowered Undo</h3>
  <ul>
  <li>
    You can now <b>undo any edit</b> that hasn't been uploaded yet: <br/>
    Quest answers, split ways, deleted places, hid quests, created notes and note comments (#1029, #2616)
  </li>
  <li>You can now look at your <b>edit history</b> and undo any recent edit in any order, not just the last one</li>
  <li>It is now shown what exactly is tagged for each edit in the edit history</li>
  <li>You can now undo deletions of places even after you have already uploaded it (#2441)</li>
  </ul>
  <h3>⛰️ Improved usability and offline mode</h3>
  <ul>
  <li>
  Immediately unlock new quests based on the one you just solved, <b>also when
  offline</b>. This has been a much requested feature but was simply not possible with the previous
  architecture (#1369, #1510, #1550, #1826, #2254, #2438, #2473, #2474, #2676, #2743...)
  </li>
  <li>
  The above also works when splitting a way. The quests for the split segments will be shown
  immediately
  </li>
  <li>When you scan for quests manually, no cache will be used: You always get the most fresh data (#2554)</li>
  <li>When unable to fully upload pictures attached to a note due to a bad connection, the app will try again later (#2102)</li>
  </ul>
  <h3>New Quests</h3>
  <ul>
  <li><i>Is this sport field lit?</i> (#2639, #2737), by @eginhard</li>
  <li><i>What type of barrier is this?</i> (#2739, #2753), by @matkoniecz</li>
  <li><i>What type of stile is this?</i> (#2749, #2766), by @matkoniecz</li>
  <li><i>What type of shop is this?</i> (#2391, #2759), by @matkoniecz</li>
  <li>Italy only: <i>Which type of police station is this?</i> (#2456, #2675), by @naposm</li>
  </ul>
  <h3>Other Enhancements</h3>
  <ul>
  <li>Enable answering that the cycleway is displayed separately on the map (#2525, #2665)</li>
  <li>Some wording improvements (#2728), by @matkoniecz, (#2806, #2814) by @peternewman, (#2685)</li>
  <li>Enhance maxspeed collection for the United Kingdom (#2745, #2748, #2750, #2811), by @arrival-spring</li>
  <li>Periodically ask for the existence of bbq spots, ticket validators, emergency life rings and emergency phones (#2679), thanks @arrival-spring</li>
  <li>Also ask for the surface of service roads as long as they are not driveways (#2783)</li>
  <li>Slightly more performant compass</li>
  </ul>
  <h3>Fixes</h3>
  <ul>
  <li>On tablets, correctly restore the position when reentering the app (#2659, #2311), thanks @matteblair</li>
  <li>Sometimes the link collection wouldn't show (#2801, #2809), by @tapetis</li>
  <li>The keyboard would not appear in the opening hours times dialog (#2799, #2820), by @tapetis</li>
  <li>The flags in the statics view were sometimes wrongly stretched (#2819, #2831), by @tapetis </li>
  <li>The time picker for the postbox collection times now follows the system settings (#2807)</li>
  <li>Fix crash on trying to display opening hours that range to 24:00 (#2830)</li>
  <li>...and other fixes and enhancements (#2521, #1821, #2628, #2691, #2715, #2728...)</li>
  <li>Remove checking existence of grit bins, as many are removed for summer and not marked as seasonal :( (#2726), by @matkoniecz</li>
  </ul>

<hr /><em>This discussion was created from the release <a href='https://github.com/streetcomplete/StreetComplete/releases/tag/v32.0-alpha1'>v32.0-alpha1</a>.</em>

**Comment by westnordost** [comment]
You are very welcome to help testing it! If you do, please report any issues you find back here or in the bugracker!

This version is only available as a direct APK download on GitHub, but when I publish a newer version on Google Play, it will automatically update to that version (so you won't need to remember updating).

**Comment by peternewman** [comment]
Installed via APK, should this have rest my disabled quest settings (which it definitely has done) and possibly the order (which I'm less certain of)?

**Reply by matkoniecz** [reply]
Yes, it is expected - see https://github.com/streetcomplete/StreetComplete/issues/2840#issuecomment-831205491

Quest order should survive.

**Reply by peternewman** [reply]
It probably did, I was less sure on that.

I'd suggest the changelog for this release should have a note at the top about that, as well as, and far more importantly, the warning from the top of the previous releases: https://github.com/streetcomplete/StreetComplete/releases/tag/v31.3

**Comment by lukashass** [comment]
Just installed this on Android 11. I had to uninstall the old version first. When `v31.3` was still installed the installer failed with `App not installed`.

**Reply by mnalis** [reply]
same thing happened to me; but I think it is because I had `v31.3` installed from f-droid, which uses different signing key than Google Play (and github APK). If so, that is normal behaviour (security protection so application can only be upgraded **from the same source** it was installed from). Also related: https://github.com/streetcomplete/StreetComplete/issues/2842

**Reply by westnordost** [reply]
Correct, the release from F-Droid is signed by F-Droid, the releases on GitHub and Google Play are signed by me. Installing one over the other is not possible.

So if you got StreetComplete from F-Droid, you need to uninstall it first.

**Comment by peternewman** [comment]
I had an instance once where it was showing e.g. road quests, but when you tapped on the pin it didn't show the outline of the way. It might still have been scanning. I couldn't reproduce it afterwards.

**Reply by westnordost** [reply]
Hm no, the quests are only created after all the map data plus geometry has been downloaded and persisted.

Also, quests are only created for elements that have a valid geometry.

So this is probably a UI issue that should be reproducable by interacting with it in a certain order or something. Please keep me posted if you find a way to reliably reproduce it!

**Reply by peternewman** [reply]
I think I just did a force quit and reopened it and it was fine, whatever that means, I guess that ties up with the UI thing.

**Reply by westnordost** [reply]
Hm, it did not ask you to send an email about the crash?

**Reply by peternewman** [reply]
It didn't crash, I just did a long press>App info>Force Stop to get it to reload, at which point it was behaving correctly.

**Reply by peternewman** [reply]
Just to say I've had the same thing again. I resolved it again with another Force Stop. Different location this time, but after the relaunch it again showed the geometry for those roads again (it was doing the same thing for two this time). Again no obvious reproduction.

**Reply by westnordost** [reply]
I created an issue: #2862

**Comment by peternewman** [comment]
I added the building levels (but not roof levels) to a building (as it was down the end of a street so I couldn't be certain if it had dormers or Velux windows. It then offered the house number quest, which was all good (although I hadn't remembered seeing it for the other buildings). However when I manually scanned quests it then showed the roof type quest. It didn't seem to immediately update when I changed the selected/enabled quests either.

I've not tried reproducing yet but haven't seen it since when using the alpha...

**Reply by westnordost** [reply]
auto-sync on or off?

**Reply by westnordost** [reply]
How many building levels did you add?

**Reply by westnordost** [reply]
Anyway, I can't reproduce this. No roof levels given is treated the same way as 0 levels given in respect to whether the roof shape quest is shown or not.
So this means that this is not a problem specific to the roof shape quest. The one way I can think of how this quest could not have been unlocked is if the geometry of that building was invalid. Which then would be the same cause as #2862

**Reply by peternewman** [reply]
> auto-sync on or off?

My auto-sync is only on WiFi, I think I was offline at the time, so I thought all the shiny new offline features should make that irrelevant.

> How many building levels did you add?

2

> Anyway, I can't reproduce this. No roof levels given is treated the same way as 0 levels given in respect to whether the roof shape quest is shown or not.
> So this means that this is not a problem specific to the roof shape quest. The one way I can think of how this quest could not have been unlocked is if the geometry of that building was invalid. Which then would be the same cause as #2862

I'm 90% certain it was showing the correct building shape/geometry when I entered the levels quest.

I should clarify I've got quests in this order:

1. Building kind
2. Levels
3. House number

The in-place upgrade wouldn't need a re-order to get it to re-read quest order would it?

**Comment by lukashass** [comment]
I encountered some cases where the _house number_ quest seems to be blocked by _building levels_ (and _roof shape_ quests).

1. I had _building levels_ disabled and wondered why no _house number_ quest was offered.
1. I then deleted all app data and saw the _building levels_ quest.
    1. When disabling the _building levels_ quest at this point no quest is offered.
    1. When answering the _building levels_ quest, the _house number_ quest is offered.
        1. When scanning for quests at this point the _house number_ quest will be replaced by the _roof shape_ quest.


This is the case on:
- https://www.openstreetmap.org/way/450582253
- https://www.openstreetmap.org/way/494982835
- https://www.openstreetmap.org/way/494981018

On a lot of other buildings I was able to answer quests in the (expected?) order of _building type_, _house number_, _street_.

**Reply by westnordost** [reply]
So, it's not just a display issue - that building levels is displayed wrongly in front of house number - but that as long as the building levels quest is not solved, the house number one does not show up?

**Reply by lukashass** [reply]
I think so, yes. When the building levels quest is disabled there is no quest displayed on the buildings mentioned above.

Sounds like I should be able to switch between available quests. Am I missing something?

**Reply by westnordost** [reply]
You can't switch, but you could disable a quest type to see if anything is "underneath" it

**Reply by lukashass** [reply]
Ah, yes. That's what I did. There is nothing underneath.

**Reply by peternewman** [reply]
Maybe this is also what I saw in https://github.com/streetcomplete/StreetComplete/discussions/2834#discussioncomment-712225 and I'm miss-describing/miss-remembering it.

**Comment by westnordost** [comment]
alpha2 and hopefully the last alpha before the public beta is out with most issues found during alpha1 testing fixed!

https://github.com/streetcomplete/StreetComplete/releases/tag/v32.0-alpha2

**Comment by RubenKelevra** [comment]
I noticed a performance issue which leads to stuttering:

Steps to reproduce

- Zoom out
- select a quest
- complete it

Example


https://user-images.githubusercontent.com/614929/118555198-b2d65100-b762-11eb-9ccd-2b77d84ae678.mp4


The view will zoom out automatically, while the answer is processed. This leads to visual stuttering even on a fairly powerful device:

Snapdragon 835
8 Gigs of RAM
UFS2.1 Storage 
Android 11

So on less powerful devices this is probably more prominantly. 

I suggest to let the animation finish before processing the answer. This way we can "hide" the stuttering most of the time, when the user is deciding what to click next.

**Reply by westnordost** [reply]
Hm this is not new though, is it?
Just checked, the processing is not done on the main thread. So, I am not sure, but likely, the stuttering is caused by the updating of the map view (i.e. in this case, remove that one quest pin), not the processing itself.

Removing a quest pin is actually not possible in the map rendering library I use (tangram-es), so what happens instead is that **all** pins are removed and re-added (minus the one that should be removed). Nothing changed here since v31.

But yeah, maybe a mitigation would be to only do the zoom out animation after the pins have been updated

**Reply by westnordost** [reply]
Oh well, I found a small thing that could be improved: Instead of zooming out immediatly on answering the question (before processing), zoom out when the processing is done and the quest is removed as a consequence of that. This still happens at the same time as the map updates its pin, but then I think the stuttering should be less noticeable as it should now happen not in the midst of the animation but somewhere at the start of the animation.

**Reply by westnordost** [reply]
At least - I didn't try it before changing this, but now I notice no stuttering at all. Same specs as your phone only instead a Snapdragon 855
